### PR TITLE
XSH-800 Allow limiting byte size of an execution result

### PR DESCRIPTION
**Pull request by Dominik-Sauer**

### CODE REVIEW

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

**Comment by Dominik-Sauer** [comment]
ok to test

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

File: gooddata-pandas/gooddata_pandas/dataframe.py [hunk 1]
```diff
@@ -17,7 +17,7 @@
     make_pandas_index,
 )
 from gooddata_sdk import Attribute, BareExecutionResponse, ExecutionDefinition, Filter, GoodDataSdk
-from gooddata_sdk.compute.model.execution import ResultSize
+from gooddata_sdk.compute.model.execution import ResultCacheMetadataWrapper, ResultSizeBytes, ResultSizeDimensions
```
Comment (hkad98) [inline gooddata-pandas/gooddata_pandas/dataframe.py:20]: If I am not mistaken you added these classes in gooddata-sdk/gooddata_sdk/__init__.py meaning you can use `from gooddata_sdk` to import them.

File: gooddata-pandas/gooddata_pandas/result_convertor.py [hunk 2]
```diff
@@ -5,7 +5,7 @@
 from attrs import define, field, frozen
 
 from gooddata_sdk import BareExecutionResponse, ExecutionResult
-from gooddata_sdk.compute.model.execution import ResultSize
+from gooddata_sdk.compute.model.execution import ResultCacheMetadataWrapper, ResultSizeBytes, ResultSizeDimensions
```
Comment (hkad98) [inline gooddata-pandas/gooddata_pandas/result_convertor.py:8]: If I am not mistaken you added these classes in gooddata-sdk/gooddata_sdk/init.py meaning you can use from gooddata_sdk to import them.

File: gooddata-pandas/tests/dataframe/test_dataframe_for_exec_def.py [hunk 3]
```diff
@@ -6,7 +6,7 @@
 
 from gooddata_pandas import DataFrameFactory
 from gooddata_sdk import Attribute, ExecutionDefinition, ObjId, SimpleMetric, TotalDefinition, TotalDimension
-from gooddata_sdk.compute.model.execution import ResultSizeLimitsExceeded
+from gooddata_sdk.compute.model.execution import ResultSizeBytesLimitExceeded, ResultSizeDimensionsLimitsExceeded
```
Comment (hkad98) [inline gooddata-pandas/tests/dataframe/test_dataframe_for_exec_def.py:9]: If I am not mistaken you added these classes in gooddata-sdk/gooddata_sdk/init.py meaning you can use from gooddata_sdk to import them.

**Review by hkad98 — CHANGES_REQUESTED** [review]

**Comment by hkad98** [comment]
@pcerny, could you please have a look at this PR as well? Thanks.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

File: gooddata-pandas/gooddata_pandas/dataframe.py (hunk 1 above)
Comment (Dominik-Sauer) [reply gooddata-pandas/gooddata_pandas/dataframe.py]: fixed

File: gooddata-pandas/gooddata_pandas/result_convertor.py (hunk 2 above)
Comment (Dominik-Sauer) [reply gooddata-pandas/gooddata_pandas/result_convertor.py]: fixed

File: gooddata-pandas/tests/dataframe/test_dataframe_for_exec_def.py (hunk 3 above)
Comment (Dominik-Sauer) [reply gooddata-pandas/tests/dataframe/test_dataframe_for_exec_def.py]: fixed

File: gooddata-sdk/gooddata_sdk/compute/model/execution.py [hunk 4]
```diff
@@ -128,13 +130,19 @@ def as_api_model(self) -> models.AfmExecution:
         return models.AfmExecution(execution=execution, result_spec=result_spec)
 
 
-ResultSize = Tuple[Optional[int], ...]
+ResultSizeDimensions = Tuple[Optional[int], ...]
+ResultSizeBytes = int
```
Comment (pcerny) [inline gooddata-sdk/gooddata_sdk/compute/model/execution.py:134]: Is there any benefit from having this alias in place? It holds the same information as variable name so it seems to be redundant. I also do not expect data type will be changed in the future for `ResultSizeBytes`

**Review by pcerny — CHANGES_REQUESTED** [review]

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

File: gooddata-pandas/gooddata_pandas/dataframe.py [hunk 5]
```diff
@@ -229,11 +237,20 @@ def for_insight(self, insight_id: str, auto_index: bool = True) -> pandas.DataFr
...
+        """
+        return self._sdk.compute.retrieve_result_cache_metadata(workspace_id=self._workspace_id, result_id=result_id)
+
     def for_exec_def(
         self,
         exec_def: ExecutionDefinition,
         label_overrides: LabelOverrides = None,
-        result_size_limits: ResultSize = (),
+        result_size_dimensions_limits: ResultSizeDimensions = (),
+        result_size_bytes_limit: ResultSizeBytes = None,
```
[the remark this answers is not in the document]
Comment (Dominik-Sauer) [reply gooddata-pandas/gooddata_pandas/dataframe.py]: fixed

File: gooddata-pandas/gooddata_pandas/dataframe.py [hunk 6]
```diff
@@ -281,7 +303,9 @@ def for_exec_result_id(
         self,
         result_id: str,
         label_overrides: LabelOverrides = None,
-        result_size_limits: ResultSize = (),
+        result_cache_metadata: ResultCacheMetadataWrapper = None,
```
[the remark this answers is not in the document]
Comment (Dominik-Sauer) [reply gooddata-pandas/gooddata_pandas/dataframe.py]: fixed

File: gooddata-pandas/gooddata_pandas/dataframe.py [hunk 7]
```diff
@@ -281,7 +303,9 @@ def for_exec_result_id(
...
+        result_size_dimensions_limits: ResultSizeDimensions = (),
+        result_size_bytes_limit: ResultSizeBytes = None,
```
[the remark this answers is not in the document]
Comment (Dominik-Sauer) [reply gooddata-pandas/gooddata_pandas/dataframe.py]: fixed

File: gooddata-pandas/gooddata_pandas/result_convertor.py [hunk 8]
```diff
@@ -125,19 +130,21 @@ def result(self) -> _DataWithHeaders:
...
 
 
-def _extract_all_result_data(
-    response: BareExecutionResponse,
-    result_size_limits: ResultSize,
+def _read_complete_execution_result(
+    execution_response: BareExecutionResponse,
+    result_cache_metadata: ResultCacheMetadataWrapper,
+    result_size_dimensions_limits: ResultSizeDimensions,
+    result_size_bytes_limit: ResultSizeBytes = None,
```
[the remark this answers is not in the document]
Comment (Dominik-Sauer) [reply gooddata-pandas/gooddata_pandas/result_convertor.py]: fixed

File: gooddata-pandas/gooddata_pandas/result_convertor.py [hunk 9]
```diff
@@ -300,10 +309,12 @@ def _merge_grand_total_headers_into_headers(extract: _DataWithHeaders) -> Tuple[
...
 
-def convert_result_to_dataframe(
-    response: BareExecutionResponse,
+def convert_execution_response_to_dataframe(
+    execution_response: BareExecutionResponse,
+    result_cache_metadata: ResultCacheMetadataWrapper,
     label_overrides: LabelOverrides,
-    result_size_limits: ResultSize,
+    result_size_dimensions_limits: ResultSizeDimensions,
+    result_size_bytes_limit: ResultSizeBytes = None,
```
[the remark this answers is not in the document]
Comment (Dominik-Sauer) [reply gooddata-pandas/gooddata_pandas/result_convertor.py]: fixed

File: gooddata-sdk/gooddata_sdk/compute/model/execution.py [hunk 10]
```diff
@@ -327,6 +335,44 @@ def __repr__(self) -> str:
...
+
+    @property
+    def result_size(self) -> int:
+        return self._result_cache_metadata.result_size
+
+    @property
+    def result_spec(self) -> ResultSpec:
+        return self._result_cache_metadata.result_spec
+
+    def check_bytes_size_limit(self, result_size_bytes_limit: ResultSizeBytes = None) -> None:
```
[the remark this answers is not in the document]
Comment (Dominik-Sauer) [reply gooddata-sdk/gooddata_sdk/compute/model/execution.py]: fixed

File: gooddata-sdk/gooddata_sdk/compute/model/execution.py [hunk 11]
```diff
@@ -327,6 +335,44 @@ def __repr__(self) -> str:
...
+    def __init__(
+        self,
+        result_size_bytes_limit: ResultSizeBytes,
+        actual_result_bytes_size: ResultSizeBytes,
+    ):
+        self.result_size_bytes_limit = result_size_bytes_limit
+        self.actual_result_bytes_size = actual_result_bytes_size
+
+
+class ResultCacheMetadataWrapper:
```
[the remark this answers is not in the document]
Comment (Dominik-Sauer) [reply gooddata-sdk/gooddata_sdk/compute/model/execution.py:347]: fixed

File: gooddata-sdk/gooddata_sdk/compute/model/execution.py (hunk 4 above)
Comment (Dominik-Sauer) [reply gooddata-sdk/gooddata_sdk/compute/model/execution.py]: fixed

File: gooddata-sdk/gooddata_sdk/compute/model/execution.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (Dominik-Sauer) [reply gooddata-sdk/gooddata_sdk/compute/model/execution.py]: fixed

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

**Comment by hkad98** [comment]
@Dominik-Sauer, how did you generate clients? I also tried to generate clients, and there is no change for the dev_latest docker image. Did you generate clients against stagging?

**Comment by Dominik-Sauer** [comment]
@hkad98 I generated against my local build.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

**Review by pcerny — APPROVED** [review]

**Comment by yenkins** [comment]
### Sonar scan result


More detail, see in https://sonarqube-gate.intgdc.com/dashboard?id=gooddata-python-sdk-gate-PR152

To scan for vulnerabilities in dependencies and run unit tests (to get coverage report in sonar) please comment your PR with 'extended check sonar'.

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-python-sdk-sonarqube-scan-zuul-gate/139/ : SUCCESS in 53s
- https://checklist.intgdc.com/job/compliance/job/gooddata-python-sdk-fossa-scanning-zuul-el7/139/ : FAILURE in 12s (non-voting)
- https://checklist.intgdc.com/job/compliance/job/gooddata-python-sdk-merge-compliance-check/141/ : SUCCESS in 17s

**Pull request merged**